C#/Botkit/botbuilder-slack-adapter/SlackMessageTypeMiddleware.cs

```diff
@@ -1,22 +1,90 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
 using System.Text;
+using System.Text.RegularExpressions;
 using System.Threading.Tasks;
 
 namespace botbuilder_slack_adapter
 {
+    /// <summary>
+    /// A middleware for Botkit developers using the BotBuilder SlackAdapter class.
+    /// This middleware causes Botkit to emit more specialized events for the different types of message that Slack might send.
+    /// Responsible for classifying messages:
+    ///      * `direct_message` events are messages received through 1:1 direct messages with the bot
+    ///      * `direct_mention` events are messages that start with a mention of the bot, i.e "@mybot hello there"
+    ///      * `mention` events are messages that include a mention of the bot, but not at the start, i.e "hello there @mybot"
+    /// In addition, messages from bots and changing them to `bot_message` events. All other types of message encountered remain `message` events.
+    /// </summary>
     public class SlackMessageTypeMiddleware : MiddlewareSet
     {
         /// <summary>
         /// Not for direct use - implements the MiddlewareSet's required onTurn function used to process the event
         /// </summary>
         /// <param name="context"></param>
         /// <param name=""></param>
-        public async void OnTurn(TurnContext context, Func<object>)
+        public async void OnTurn(TurnContext context, Func<Task<object>> next)
         {
+            if (context.Activity.Type == "message" && context.Activity.ChannelData != null)
+            {
+                var adapter = context.Adapter as SlackAdapter;
 
+                string bot_user_id = await adapter.GetBotUserByTeam(context.Activity);
+                var mentionSyntax = "<@" + bot_user_id + "(\\|.*?)?>";
+                var mention = new Regex(mentionSyntax, RegexOptions.IgnoreCase);
+                var direct_mention = new Regex('^' + mentionSyntax, RegexOptions.IgnoreCase);
+
+                // is this a DM, a mention, or just ambient messages passing through?
+                if ((context.Activity.ChannelData as dynamic)?.channel_type == "im")
+                {
+                    (context.Activity.ChannelData as dynamic).botkitEventType = "direct_message";
+
+                    // strip any potential leading @mention
+                    Regex.Replace(
+                        Regex.Replace(
+                            Regex.Replace(
+                                Regex.Replace(context.Activity.Text, direct_mention.ToString(), ""), 
+                                @"/ ^\s +/", ""), 
+                            @"/ ^:\s +/", ""), 
+                        @"/ ^\s +/", "");
+                }
+                else if (!string.IsNullOrEmpty(bot_user_id) && !string.IsNullOrEmpty(context.Activity.Text) && context.Activity.Text.Equals(direct_mention))
+                {
+                    (context.Activity.ChannelData as dynamic).botkitEventType = "direct_mention";
```

Comment:
Do not forget to update these hardcoded properties.
In fact, depending on how it is used, we could use an `enum` to assign this value